cairo1-run/src/main.rs

```diff
@@ -352,8 +359,67 @@ fn run(args: impl Iterator<Item = String>) -> Result<Vec<MaybeRelocatable>, Erro
         }
     }
 
+    // Set stop pointers for builtins so we can obtain the air public input
+    if args.air_public_input.is_some() {
+        // Cairo 1 programs have other return values aside from the used builtin's final pointers, so we need to hand-pick them
+        let ret_types_sizes = main_func
+            .signature
+            .ret_types
+            .iter()
+            .map(|id| type_sizes.get(id).cloned().unwrap_or_default());
+        let ret_types_and_sizes = main_func
+            .signature
+            .ret_types
+            .iter()
+            .zip(ret_types_sizes.clone());
+
+        let full_ret_types_size: i16 = ret_types_sizes.sum();
+        let mut stack_pointer = (vm.get_ap() - (full_ret_types_size as usize).saturating_sub(1))
+            .map_err(VirtualMachineError::Math)?;
+
+        // Calculate the stack_ptr for each return builtin in the return values
+        let mut builtin_name_to_stack_pointer = HashMap::new();
+        for (id, size) in ret_types_and_sizes {
+            if let Some(ref name) = id.debug_name {
+                let builtin_name = match &*name.to_string() {
+                    "RangeCheck" => RANGE_CHECK_BUILTIN_NAME,
+                    "Poseidon" => POSEIDON_BUILTIN_NAME,
+                    "EcOp" => EC_OP_BUILTIN_NAME,
+                    "Bitwise" => BITWISE_BUILTIN_NAME,
+                    "Pedersen" => HASH_BUILTIN_NAME,
+                    "Output" => OUTPUT_BUILTIN_NAME,
+                    "Ecdsa" => SIGNATURE_BUILTIN_NAME,
+                    _ => {
+                        stack_pointer.offset += size as usize;
+                        continue;
+                    }
+                };
+                builtin_name_to_stack_pointer.insert(builtin_name, stack_pointer);
+            }
+            stack_pointer.offset += size as usize;
+        }
+        // Set stop pointer for each builtin
+        for builtin in vm.builtin_runners.iter_mut() {
+            builtin.final_stack(
+                &vm.segments,
+                builtin_name_to_stack_pointer
+                    .get(builtin.name())
+                    .cloned()
+                    .unwrap_or_default(),
+            )?;
+        }
```

Comment:
Can't `finalize_segments` do this so we avoid exposing `VirtualMachine::builtin_runners` and `VirtualMachine::segments`?

Reply:
In cairo 0, the builtin's stop pointers can be found contiguously at the end of the execution segment. Therefore we can take the current value of ap as stack_pointer and then iterate over the builtin runners calling `final_stack` (which will fetch the builtin's stop_ptr from memory[stack_ptr - 1]) on each of them, while advancing the stack_ptr by one each iteration to set all builtin's stop_ptr to the correct value. 
In cairo 1 this doesn't happen, the builtin's stop_ptr are mixed up with the other return values so we can't iterate contiguously over a memory range to fetch the stop pointers. The logic above this code calculates an individual stack_ptr for each builtin, therefore we need to call each builtin's `final_stack` method independently, without being able to rely on `read_return_values` or `get_builtins_final_stack` like we do with cairo 0 execution.

Reply:
Maybe a new method in the runner can do it?

Reply:
It is kind of an specific behaviour to have as part of the public api, but I think we could add it to the vm as a `docs_hidden` method